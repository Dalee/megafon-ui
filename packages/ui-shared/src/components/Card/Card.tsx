import * as React from 'react';
import './Card.less';
import { cnCreate, Header, Paragraph, Button, TextLink, Link } from '@megafon/ui-core';
import PropTypes from 'prop-types';

interface IButton {
    title: string;
    href: string;
}

interface ILink {
    title: string;
    href?: string;
}

export const ObjectFit = {
    FILL: 'fill',
    CONTAIN: 'contain',
} as const;

type ObjectFitType = typeof ObjectFit[keyof typeof ObjectFit];

interface ICard {
    /** Изображение в карточке */
    imageSrc?: string;
    /** Иконка в карточке */
    svgSrc?: React.ReactNode;
    /** Заголовок карточки */
    title: string;
    /** Текст карточки */
    text: string;
    /** Данные для кнопки */
    button?: IButton;
    /** Данные для ссылки */
    link?: ILink;
    /** Расположение кнопки/ссылки по левой стороне */
    isLeftHAlign?: boolean;
    /** Ссылка для всей карточки */
    href?: string;
    /** Режим позиционирования изображения */
    objectFit?: ObjectFitType;
}

const cn = cnCreate('mfui-beta-card');
const Card: React.FC<ICard> = ({
    imageSrc,
    svgSrc,
    title,
    text,
    button,
    link,
    isLeftHAlign = false,
    href,
    objectFit = 'fill',
}) => {
    const isAlignAvailable = !button || !link;
    const isCardLink = !!href;
    const isRenderBtn = !!button && !isCardLink;
    const Element = href ? Link : 'div';

    const renderImage = React.useCallback(() => {
        switch (true) {
            case !!imageSrc: {
                return (
                    <div className={cn('pic-wrapper', { 'object-fit': objectFit, 'img': true })}>
                        <img className={cn('img')} src={imageSrc} />
                    </div>
                );
            }

            case !!svgSrc: {
                return (
                    <div className={cn('pic-wrapper')}>
                        {svgSrc}
                    </div>
                );
            }

            default:
                return null;
        }
    }, [imageSrc, svgSrc, objectFit]);

    const renderLink = React.useCallback(({ href: linkHref, title: linkTitle }) => {
        const isFakeLink = !linkHref;

        if (isFakeLink || isCardLink) {
            return <span className={cn('fake-link')}>{linkTitle}</span>;
        }

        return (
            <TextLink className={cn('link')} href={linkHref}>{linkTitle}</TextLink>
        );
    }, [isCardLink]);

    const renderBtn = React.useCallback(({ href: btnHref, title: btnTitle}) => (
        <Button className={cn('button')} href={btnHref}>{btnTitle}</Button>
    ), []);

    return (
        <div className={cn('', { 'href': !!href })}>
            <Element href={href}>
                <div className={cn('inner')}>
                    {renderImage()}
                    <Header as="h3" className={cn('title')}>{title}</Header>
                    <Paragraph hasMargin={false}>{text}</Paragraph>
                    <div className={cn('btns-wrapper', { 'left-align': isAlignAvailable && isLeftHAlign })}>
                        {isRenderBtn && renderBtn(button)}
                        {link && renderLink(link)}
                    </div>
                </div>
            </Element>
        </div>
    );
};

Card.propTypes = {
    imageSrc: PropTypes.string,
    svgSrc: PropTypes.node,
    title: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    button: PropTypes.shape({
        title: PropTypes.string.isRequired,
        href: PropTypes.string.isRequired,
    }),
    link: PropTypes.shape({
        title: PropTypes.string.isRequired,
        href: PropTypes.string,
    }),
    isLeftHAlign: PropTypes.bool,
    href: PropTypes.string,
    objectFit: PropTypes.oneOf(Object.values(ObjectFit)),
};

export default Card;
