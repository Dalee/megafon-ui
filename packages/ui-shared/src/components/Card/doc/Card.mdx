---
name: Card
menu: Конструктор
route: components/card
---

import Example from './Card.example.mdx';
import Props from './Card.props.mdx';

# Card

<ImportExample name="Card" />

<DoczTabs>
    <Example />
    <Props />
</DoczTabs>
