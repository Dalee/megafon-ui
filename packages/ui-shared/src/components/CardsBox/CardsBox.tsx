import * as React from 'react';
import * as PropTypes from 'prop-types';
import throttle from 'lodash.throttle';
import { cnCreate, Grid, GridColumn, Carousel, breakpoints } from '@megafon/ui-core';
import { ICard } from '../Card/Card';
import { SlidesSettingsType } from '@megafon/ui-core/dist/es/components/Carousel/Carousel';

const MAX_CARDS_COUNT_IN_GRID_ON_MOBILE = 2;
const THROTTLE_TIME = 750;

const SlidesSettings: SlidesSettingsType = {
    [breakpoints.mobileSmallStart]: {
        slidesPerView: 1,
        spaceBetween: 16,
    },
    [breakpoints.mobileBigStart]: {
        slidesPerView: 2,
        spaceBetween: 20,
    },
};

interface ICardsBox {
    children: Array<React.ReactElement<ICard>> | React.ReactElement<ICard>;
}

const cn = cnCreate('mfui-beta-cards-box');
const CardsBox: React.FC<ICardsBox> = ({ children }) => {
    const [isMobile, setIsMobile] = React.useState(false);
    const itemsCount = React.Children.count(children);
    const isRenderCarousel = isMobile && itemsCount > MAX_CARDS_COUNT_IN_GRID_ON_MOBILE;

    const renderGrid = React.useCallback(() => (
        <Grid guttersBottom="medium" guttersLeft="medium">
            {
                React.Children.map(children, child => (
                    <GridColumn all="4" mobile="12">
                        {child}
                    </GridColumn>
                ))
            }
        </Grid>
    ), []);

    const renderCarousel = React.useCallback(() => (
        <Carousel slidesSettings={SlidesSettings}>{children}</Carousel>
    ), []);

    React.useEffect(() => {
        const resizeHandler = () =>
            window.innerWidth <= breakpoints.mobileBigEnd ? setIsMobile(true) : setIsMobile(false);
        const resizeHandlerThrottled = throttle(resizeHandler, THROTTLE_TIME);

        resizeHandler();
        window.addEventListener('resize', resizeHandlerThrottled);

        return () => {
            window.removeEventListener('resize', resizeHandlerThrottled);
        };
    }, []);

    return (
        <div className={cn()}>
            {isRenderCarousel ? renderCarousel() : renderGrid()}
        </div>
    );
};

CardsBox.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.element.isRequired),
        PropTypes.element,
    ]).isRequired,
};

export default CardsBox;
