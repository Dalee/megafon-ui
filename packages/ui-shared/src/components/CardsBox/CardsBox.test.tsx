import * as React from 'react';
import { shallow } from 'enzyme';
import CardsBox from './CardsBox';
import Card, { ICard } from '../Card/Card';

const cardProps: ICard = {
    title: 'title',
    text: 'text',
    button: {
        title: 'button-title',
        href: 'button-href',
    },
    link: {
        title: 'href-title',
        href: 'href',
    },
};

describe('CardsBox', () => {
    it('render component', () => {
        const wrapper = shallow(<CardsBox><Card {...cardProps} /></CardsBox>);
        expect(wrapper).toMatchSnapshot();
    });
});
