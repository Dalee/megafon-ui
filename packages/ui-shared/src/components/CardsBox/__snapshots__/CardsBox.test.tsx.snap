// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsBox render component 1`] = `
<div
  className="mfui-beta-cards-box"
>
  <Grid
    guttersBottom="medium"
    guttersLeft="medium"
    multiRow={true}
  >
    <GridColumn
      all="4"
      flex={false}
      grow={false}
      key=".0"
      mobile="12"
      orderAll="0"
      orderDesktop="0"
      orderMobile="0"
      orderTablet="0"
      orderWide="0"
    >
      <Card
        button={
          Object {
            "href": "button-href",
            "title": "button-title",
          }
        }
        link={
          Object {
            "href": "href",
            "title": "href-title",
          }
        }
        text="text"
        title="title"
      />
    </GridColumn>
  </Grid>
</div>
`;
